cmd/kubelet/app/server.go

```diff
@@ -806,8 +807,21 @@ func buildKubeletClientConfig(s *options.KubeletServer, nodeName types.NodeName)
 	}
 
 	kubeClientConfigOverrides(s, clientConfig)
+	closeAllConns, err := updateDialer(clientConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	return clientConfig, closeAllConns, nil
```

Comment:
if we require buildKubeletClientConfig to always give us back a closeall function, that would have caught this earlier. Can we add that check here and error early, rather than assigning `nil` to `OnHeartbeatFailure`?

https://github.com/kubernetes/kubernetes/blob/d5876954e15678eb643456d25e7e150d04706113/cmd/kubelet/app/server.go#L566-L570

Reply:
IMHO, this check seems to be a little bit tedious, since if err is `nil`,  `closeAllConns` must not be nil :)

Reply:
the logic in buildKubeletClientConfig is sufficiently complex that if it changed, and started returning a nil closeAllConns, no clear failures would occur, it would pass CI, and make it back into a release as another regression. A nil check before assigning to OnHeartbeatFailure would make the error immediately clear.

Reply:
done, thanks for the explanation.